tests/test_base_device.py

```diff
@@ -0,0 +1,1056 @@
+                    qml.var(qml.PauliX(0)),
+                ],
+                (0, 1, 1),
+            ),
+        ],
+    )
+    def test_process_estimator_job(self, measurements, expectation):
+        """for variance and for expval and for a combination"""
+
+        # make PennyLane circuit
+        qs = QuantumScript([], measurements=measurements)
+
+        # convert to Qiskit circuit information
+        qcirc = circuit_to_qiskit(qs, register_size=qs.num_wires, diagonalize=False, measure=False)
+        pauli_observables = [mp_to_pauli(mp, qs.num_wires) for mp in qs.measurements]
+
+        # run on simulator via Estimator
+        estimator = Estimator(backend=backend)
+        result = estimator.run([qcirc] * len(pauli_observables), pauli_observables).result()
+
+        # confirm that the result is as expected - if the test fails at this point, its because the
+        # Qiskit result format has changed
+        assert isinstance(result, EstimatorResult)
+
+        assert isinstance(result.values, np.ndarray)
+        assert result.values.size == len(qs.measurements)
+
+        assert isinstance(result.metadata, list)
+        assert len(result.metadata) == len(qs.measurements)
+
+        for data in result.metadata:
+            assert isinstance(data, dict)
+            assert list(data.keys()) == ["variance", "shots"]
+
+        processed_result = QiskitDevice2._process_estimator_job(qs.measurements, result)
+        assert isinstance(processed_result, tuple)
+        assert np.allclose(processed_result, expectation, atol=0.05)
+
+    @pytest.mark.parametrize("num_wires", [1, 3, 5])
+    @pytest.mark.parametrize("num_shots", [50, 100])
+    def test_generate_samples(self, num_wires, num_shots):
+        qs = QuantumScript([], measurements=[qml.expval(qml.PauliX(0))])
+
+        qcirc = circuit_to_qiskit(qs, register_size=num_wires, diagonalize=True, measure=True)
+        compiled_circuits = test_dev.compile_circuits([qcirc])
+
+        # Send circuits to the cloud for execution by the circuit-runner program
+        job = test_dev.service.run(
+            program_id="circuit-runner",
+            options={"backend": backend.name},
+            inputs={"circuits": compiled_circuits, "shots": num_shots},
+        )
+
+        test_dev._current_job = job.result(decoder=RunnerResult)
+
+        samples = test_dev.generate_samples()
+
+        assert len(samples) == num_shots
+        assert len(samples[0]) == num_wires
+
+        # we expect the samples to be orderd such that q0 has a 50% chance
+        # of being excited, and everything else is in the ground state
+        exp_res0 = np.zeros(num_wires)
+        exp_res1 = np.zeros(num_wires)
+        exp_res1[0] = 1
+
+        # the two expected results are in samples
+        assert exp_res1 in samples
+        assert exp_res0 in samples
+
+        # nothing else is in samples
+        assert [s for s in samples if not s in np.array([exp_res0, exp_res1])] == []
+
```

Comment:
Not blocking, but for good measure, it might be nice to add a few more lines in these tape shots tests, so that after each call with `circuit(shots=[5])`, you call it again with no shots and confirm it returns to the default rather than being permanently modified.

Also let's add a short docstring to each - it makes things much nicer for people coming back to troubleshoot if we break something a year from now 😅 Just something like "Tests that if shots are passed to the circuit at execution, they override the default device shots for that execution for the [insert which type of execute function here] function."

Reply:
> Not blocking, but for good measure, it might be nice to add a few more lines in these tape shots tests, so that after each call with `circuit(shots=[5])`, you call it again with no shots and confirm it returns to the default rather than being permanently modified.
I missed that. I think I did this while manually testing it myself but then just didn't do it for the actual tests for some reason. Should be a quick fix though.
 
> Also let's add a short docstring to each - it makes things much nicer for people coming back to troubleshoot if we break something a year from now 😅 Just something like "Tests that if shots are passed to the circuit at execution, they override the default device shots for that execution for the [insert which type of execute function here] function."
Will add docstrings! Forgot 😅